### ISSUE TITLE
Reindex from remote can break mappings

### ISSUE THREAD

**Issue by aleph-zero** [elastic/elasticsearch#21033]
**Elasticsearch version**:
2.4 and 5.0-RC1

**Plugins installed**: []
On 2.4 cluster:
```
tiebreaker-0000000082 analysis-icu         2.4.1                j  
tiebreaker-0000000082 analysis-kuromoji    2.4.1                j  
tiebreaker-0000000082 analysis-smartcn     2.4.1                j  
tiebreaker-0000000082 cloud-aws            2.4.1                j  
tiebreaker-0000000082 found-elasticsearch  1.2.23-2.4.1-c67dc32 j  
tiebreaker-0000000082 found-license-plugin 2.4.1                j  
tiebreaker-0000000082 graph                2.4.1                j  
tiebreaker-0000000082 marvel-agent         2.4.1                j  
tiebreaker-0000000082 shield               2.4.1                j  
tiebreaker-0000000082 watcher              2.4.1                j  
instance-0000000086   analysis-icu         2.4.1                j  
instance-0000000086   analysis-kuromoji    2.4.1                j  
instance-0000000086   analysis-smartcn     2.4.1                j  
instance-0000000086   cloud-aws            2.4.1                j  
instance-0000000086   found-elasticsearch  1.2.23-2.4.1-c67dc32 j  
instance-0000000086   found-license-plugin 2.4.1                j  
instance-0000000086   graph                2.4.1                j  
instance-0000000086   marvel-agent         2.4.1                j  
instance-0000000086   shield               2.4.1                j  
instance-0000000086   watcher              2.4.1                j  
instance-0000000083   analysis-icu         2.4.1                j  
instance-0000000083   analysis-kuromoji    2.4.1                j  
instance-0000000083   analysis-smartcn     2.4.1                j  
instance-0000000083   cloud-aws            2.4.1                j  
instance-0000000083   found-elasticsearch  1.2.23-2.4.1-c67dc32 j  
instance-0000000083   found-license-plugin 2.4.1                j  
instance-0000000083   graph                2.4.1                j  
instance-0000000083   marvel-agent         2.4.1                j  
instance-0000000083   shield               2.4.1                j  
instance-0000000083   watcher              2.4.1                j  
```

On 5.0-RC1 cluster:
```
tiebreaker-0000000003 found-elasticsearch 5.0.0-rc1
tiebreaker-0000000003 repository-s3       5.0.0-rc1
tiebreaker-0000000003 x-pack              5.0.0-rc1
instance-0000000005   found-elasticsearch 5.0.0-rc1
instance-0000000005   repository-s3       5.0.0-rc1
instance-0000000005   x-pack              5.0.0-rc1
instance-0000000004   found-elasticsearch 5.0.0-rc1
instance-0000000004   repository-s3       5.0.0-rc1
instance-0000000004   x-pack              5.0.0-rc1
```

**JVM version**:
Unknown. Both clusters are running on Elastic Cloud.

**OS version**:
Unknown. Both clusters are running on Elastic Cloud.

**Description of the problem including expected versus actual behavior**:
When reindexing from a remote 2.4 cluster on to a new 5.0-RC1 cluster, some field mappings were incorrectly set on the new index. Specifically, I noticed that a field of type ```geo_point``` was created as type ```text``` on the new index on the 5.0-RC1 cluster.

Look at the **location** field on the mappings.

Index mapping on 2.4:
```
{
  "ncedc-earthquakes": {
    "mappings": {
      "earthquake": {
        "properties": {
          "@timestamp": {
            "type": "date",
            "format": "strict_date_optional_time||epoch_millis"
          },
          "@version": {
            "type": "string",
            "index": "not_analyzed"
          },
          "depth": {
            "type": "double"
          },
          "dmin": {
            "type": "double"
          },
          "event_id": {
            "type": "string",
            "index": "not_analyzed"
          },
          "gap": {
            "type": "double"
          },
          "host": {
            "type": "string",
            "index": "not_analyzed"
          },
          "location": {
            "type": "geo_point"
          },
          "mag": {
            "type": "double"
          },
          "magType": {
            "type": "string",
            "index": "not_analyzed"
          },
          "message": {
            "type": "string",
            "index": "not_analyzed"
          },
          "nst": {
            "type": "string",
            "index": "not_analyzed"
          },
          "rms": {
            "type": "double"
          },
          "source": {
            "type": "string",
            "index": "not_analyzed"
          },
          "tags": {
            "type": "string",
            "index": "not_analyzed"
          },
          "timestamp": {
            "type": "string",
            "index": "not_analyzed"
          },
          "type": {
            "type": "string",
            "index": "not_analyzed"
          }
        }
      }
    }
  }
}
```

Index mapping on 5.0-RC1 (automatically created during reindex-from-remote):
```
{
  "ncedc-earthquakes": {
    "mappings": {
      "earthquake": {
        "properties": {
          "@timestamp": {
            "type": "date"
          },
          "@version": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "depth": {
            "type": "float"
          },
          "dmin": {
            "type": "float"
          },
          "event_id": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "gap": {
            "type": "float"
          },
          "host": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "location": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "mag": {
            "type": "float"
          },
          "magType": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "message": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "nst": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "rms": {
            "type": "float"
          },
          "source": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "tags": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "timestamp": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          },
          "type": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword",
                "ignore_above": 256
              }
            }
          }
        }
      }
    }
  }
}
```

**Steps to reproduce**:
 1. Create an index with a ```geo_point``` field in a 2.4 cluster.
 2. Reindex from the 2.4 cluster into the 5.0-RC1 cluster.
 3. Look at the mappings for the ```geo_point``` field.

**Comment by nik9000** [comment]
I'll have a closer look in the morning. For now you should be able to work
around it by creating the mapping before starting the reindex. That is
usually the right way to use reindex anyway. If there is a way to make it
work automatically I'll fix it though.

**Comment by jpountz** [comment]
Hmm I'm confused about why reindex from remote is concerned here. Isn't the issue just that Elasticsearch never creates dynamic `geo_point` mappings automatically? Which we do not do since it is hard to know with good confidence whether a string is a geo hash or a random keyword, or whether a double array containing 2 values is a geo point or a double array.

**Comment by nik9000** [comment]
If we never do then, yeah, this doesn't have to do with reindex from remote. That is what I was going to look in to this morning. I thought there might be a parameter that reindex could set that'd get the field in a more obvious way. I was tired so I didn't realize how strange that'd be to have in the first place.

In that case I think we should just close this. We've recently fixed the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/5.0/docs-reindex.html) so they are more clear that reindex doesn't set up the destination index. We have https://github.com/elastic/elasticsearch/pull/20576 where we looked at making reindex not start if the destination index doesn't exist. The implementation is fairly horrible because Elasticsearch doesn't have the right hooks in the right spots and adding them doesn't seem worth it....
